Src/Common/CorrelationIdLookupHelper.cs

```diff
@@ -0,0 +1,158 @@
+        /// </summary>
+        private const int MAXSIZE = 100;
+
+        private const int GET_APP_ID_TIMEOUT = 2000; // milliseconds
+
+        private const string CORRELATION_ID_FORMAT = "aid-v1:{0}";
+
+        private const string APPID_QUERY_API_RELATIVE_URI_FORMAT = "api/profiles/{0}/appId";
+
+        private static ConcurrentDictionary<string, string> knownCorrelationIds = new ConcurrentDictionary<string, string>();
+
+        private static Func<string, string, Task<string>> provideAppId = FetchAppIdFromService;
+
+        /// <summary>
+        /// This is a test hook. Use this to provide your own test implementation for fetching the appId.
+        /// </summary>
+        /// <param name="appIdProviderMethod">The delegate to be called to fetch the appId</param>
+        public static void OverrideAppIdProvider(Func<string, string, Task<string>> appIdProviderMethod)
+        {
+            provideAppId = appIdProviderMethod;
+        }
+
+        /// <summary>
+        /// Retrieves the correlation id corresponding to a given instrumentation key.
+        /// </summary>
+        /// <param name="instrumentationKey">Instrumentation key string.</param>
+        /// <param name="breezeEndpointAddress">Breeze endpoint to query against.</param>
+        /// <param name="correlationId">AppId corresponding to the provided instrumentation key</param>
+        /// <returns>true if correlationId was successfully retrieved; false otherwise.</returns>
+        public static bool TryGetXComponentCorrelationId(string instrumentationKey, string breezeEndpointAddress, out string correlationId)
+        {
+            if (string.IsNullOrEmpty(instrumentationKey))
+            {
+                throw new ArgumentNullException("instrumentationKey");
+            }
+
+            var found = knownCorrelationIds.TryGetValue(instrumentationKey, out correlationId);
+
+            if (found)
+            {
+                return true;
+            }
+            else
+            {
+
+                if (breezeEndpointAddress == null)
+                {
+                    throw new ArgumentNullException("breezeEndpointAddress");
+                }
+
+                // Simplistic cleanup to guard against this becoming a memory hog.
+                if (knownCorrelationIds.Keys.Count >= MAXSIZE)
+                {
+                    knownCorrelationIds.Clear();
+                }
+
+                try
+                {
+                    // Todo: When this fails, say in the vortex endpoint case, ProfileQueryEndpont is not provided, it may perpetually keep failing.
+                    // We can possibly make it more robust by having an exponential backoff on making a call to prod endpoint. Or store failure and never query again.
+                    // Is that worth the effort?
+                    
+                    // We wait for 2 seconds to retrieve the appId. If retrieved during that time, we return success setting the correlation id.
```

Comment:
Why do we still wait for even 2 seconds here?

Even though the current response is expected to be under 500 ms, it may still screw up the operation API statistics if it takes 100 ms without SDK code on this path.
Imagine 9 calls by 10 ms and 1 call with 500 ms leading to an average of 59 ms - 6 times slower than without AI SDK.
Looking at the statistics before/after I tried AI I would think of turning AI off due to 6x slow down on my fast dependency calls.

This comes in addition to the execution of the http call in the dependency collection context that (in case or insufficient protection on our side) can lead to another dependency collected (for our call) and unnecessary stacks (if we do not get it from the first time and our own call that's collected as dependency will initiate another call to AI Endpoint).

Also, the execution of the external communication in the user code context does not sound like a good idea in general. Can we make this code entirely asynchronous and then take the value if it was already populated?